### ISSUE TITLE
Bot selling at loss even when max_sell_loss_pct is set to very low value (0.1)

### ISSUE THREAD

**Issue by code4passion** [DeviaVir/zenbot#941]
I see that bot sometimes sells at a loss even when max_sell_loss_pct is configured to very less value.

Here  are the trades which the bot made in live mode.

```
trade3 SELL 2.7433 815.60 0.0000 6h ago ETH-USD
trade2 BUY  0.8777 802.58 0.0000 6h ago ETH-USD
trade1 BUY  1.8745 841.28 0.0000 9h ago ETH-USD
```
Strategy used : trend_ema (default strategy)

NOTE: Bot was never restarted between trades.It was left untouched.

It happened cause  the bot made 2 buy trades , the second trade trade2 was at a lower price then first trade trade1. Hence the bot sold at a price higher then the buy of trade2. It dinnot remeber that it actually bought the coins in trade1 at much higher price.

I am running 3 instances of bot with 33% capital on BTC,LCT,ETH. Is this because of that?
Why bot buys in 2 separate trade? May be it is because the available capital changed. May be other bot instant sold something . I dinot add/remove funds.

Expected behavior:
The bot should remember its past buyt trade at least in same session.

**Comment by code4passion** [comment]
I have observer other examples as well. Lets say bot buys at p1,p2,p3,p4 in 4 different trade . So it sells a price higher then min (p1,p2,p3,p4). 

Expected behavior is that it should sell at price higher then max(p1,p2,p3,p4) Or atleast  if its in profit

**Comment by code4passion** [comment]
Here is the culprit https://github.com/carlos8f/zenbot/blob/a581beb1faa16a495205f30d979832aad7f3944c/lib/engine.js#L478

**Comment by code4passion** [comment]
https://github.com/carlos8f/zenbot/blob/a581beb1faa16a495205f30d979832aad7f3944c/lib/engine.js#L478

The above code explain the behavior in the original post. But i can also see the following in my trade history. When bot does multiple buys in a row , it then sells at a price higher then min of all buys if loss protection(xax_sell_loss_pct ==0)  is enabled.

What is the reason for following behavior?

![image](https://user-images.githubusercontent.com/9897265/34193425-8702bb48-e508-11e7-9502-1792a8976c9d.png)

**Comment by pursehouse** [comment]
do you have a fix for this? I haven't got far enough in the code to access the entire list of sales to do a max() thing yet

**Comment by pursehouse** [comment]
quick glance, I don't have time to test right now but I really wanted to fix it If someone could test/confirm it that would be awesome.. this is a replacement of the line you found @code4passion 

```

              var oldPrices = [];
              s.my_trades.forEach(function ( oldTrade ) {
                  if( oldTrade.last_buy_price > 0 ) {
                      oldPrices.push( oldTrade.last_buy_price );
                  }
              });

              // not sure if this should be here because it would prevent buys outside of the history from happening.
              // perhaps it should read from the history to populate the s.my_trades
              if( oldPrices.length == 0 ) {
                  var err = new Error('\nloss protection')
                  err.desc = 'refusing to sell at ' + fc(price) + ', no old buys found ';
                  return cb(err)
              }

              var maxOldBuy = Math.max.apply( null, oldPrices );
              var sell_loss = maxOldBuy ? (Number(price) - maxOldBuy ) / maxOldBuy * -100 : null

```

**Comment by pursehouse** [comment]
fixed... but I forsee issues with buy/sells being mixed weird. like if the sell isn't at 100%
```
              var oldPrices = [];
              s.my_trades.forEach(function ( oldTrade ) {
//                  console.log( '\n oldTrade.last_buy_price ' + oldTrade.price );
//                  console.log( '\n' );
//                  console.log( oldTrade );
                  if( oldTrade.type == 'buy' ) {
                      if( oldTrade.price > 0 ) {
                          oldPrices.push( oldTrade.price );
                      }
                  } if( oldPrices.length > 0 ) {
                      return;
                  }
              });

//              console.log( "\n trades length : " + s.my_trades.length );

              // not sure if this should be here because it would prevent buys outside of the history from happening.
              // perhaps it should read from the history to populate the s.my_trades
              if( oldPrices.length == 0 ) {
                  var err = new Error('\nloss protection')
                  err.desc = 'refusing to sell at ' + fc(price) + ', no old buys found ';
                  return cb(err)
              }

              var maxOldBuy = Math.max.apply( null, oldPrices );
              var sell_loss = maxOldBuy ? (Number(price) - maxOldBuy ) / maxOldBuy * -100 : null
```

**Comment by DeviaVir** [comment]
This issue was moved to highvelocityspace/zenbot#30

**Comment by DeviaVir** [comment]
This issue was moved to highvelocityspace/zenbot#60

**Comment by LoneWolf345** [comment]
Are you still experiencing this problem?  If not, consider closing this.
